cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -381,6 +388,24 @@ AzureOptions::MakeDataLakeServiceClient() const {
   return Status::Invalid("AzureOptions doesn't contain a valid auth configuration");
 }
 
+Result<std::string> AzureOptions::GenerateSASToken(
+    Storage::Sas::BlobSasBuilder* builder) const {
+  if (storage_shared_key_credential_) {
+    return builder->GenerateSasToken(*storage_shared_key_credential_);
+  } else {
+    // This part isn't tested. This may not work.
```

Comment:
@Tom-Newton It seems that we can't use `GetUserDelegationKey()` with account key credential. We may need to other authentication to test `GetUserDelegationKey()`. Do you know how to configure other authentication...?

(I think that we can defer this case to a separated issue.)

Reply:
I'm pretty sure its possible to create a SAS token from an account key but it probably isn't called delegated anymore.

Reply:
Yes. We can create a SAS token from an account key. The above `builder->GenerateSasToken(*storage_shared_key_credential_)` does it.

Reply:
Sorry I misunderstood the intention (I haven't found time to review this). 

Configuring other auths in testing might be a bit difficult. For Azurite I think the first challenge is configuring `https`, which I have struggled with before. Beyond that I think there are 2 viable options offered by Azure:
1. Service principal
    1. Username and password for a service account
    2. Surprisingly complicated to create
1. Azure CLI 
    1. Authenticates as your user
    2. Requires installing the Azure CLI and running `az login`, then completing the login in your browser.

Reply:
Thanks.
I tried `az login` and `ConfigureDefaultCredential()` but some operations such as `CreateDir()` was blocked...
It seems that we need to implement GH-39344 to test this part.

I'll open a new issue for this part and defer this part in this PR.

Reply:
#41315

Reply:
> I tried az login and ConfigureDefaultCredential() but some operations such as CreateDir() was blocked...
It seems that we need to implement https://github.com/apache/arrow/issues/39344 to test this part.

This seems a bit suspicious. Default credential does include CLI auth...

Reply:
Let's look at this in #41315.